File: Swap Nodes in Pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public :
    ListNode * swap(ListNode *left, ListNode *mid, ListNode *right){
        if(!right){
            return mid;
        }
        mid->next = right->next;
        right->next = mid;
        if(left) left->next = right;
        if(mid->next){
            swap(mid, mid->next, mid->next->next);
        }
        return right;
    }

    ListNode *swapPairs(ListNode *head) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        ListNode *left, *right;
        if(! head || ! head->next){
            return head;
        }
        return swap( NULL, head, head->next);
    }
};

File: python/Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        if head == None or head.next == None:
            return head
        else:
            return self.swap(head, head.next)
            
    def swap(self, left, right):
        afterRight = right.next
        right.next = left
        if afterRight != None and afterRight.next != None:
            left.next = self.swap(afterRight, afterRight.next)
        else:
            left.next = afterRight
        return right